parlai/nn/lr_scheduler.py

```diff
@@ -0,0 +1,357 @@
+            # metrics to adjust schedule
+            return
+        if 'loss' not in metrics_dict:
+            # nothing to step on, just skip
+            warn_once("LR scheduler expected to see loss metric, but didn't.")
+            return
+        self.scheduler.step(metrics_dict['loss'])
+
+
+class FixedLRScheduler(ParlAILRScheduler):
+    """
+    Scheduler that decays by a fixed multiplicative rate at each valid step.
+    """
+
+    def __init__(self, optimizer, states, hard_reset, patience, decay, warmup_updates):
+        super().__init__(optimizer, states, hard_reset, warmup_updates)
+        self.scheduler = optim.lr_scheduler.StepLR(optimizer, patience, gamma=decay)
+
+    def train_step(self, scheduler_steps):
+        pass
+
+    def valid_step(self, metrics_dict):
+        if self._is_lr_warming_up():
+            # we're not done warming up, so don't start using validation
+            # metrics to adjust schedule
+            return
+        self.scheduler.step()
+
+
+class InvSqrtLRScheduler(ParlAILRScheduler):
+    """
+    Scheduler that decays at an inverse square root rate.
+    """
+
+    def __init__(
+        self,
+        optimizer,
+        states,
+        hard_reset,
+        patience,
+        decay,
+        warmup_updates,
+        max_lr_steps,
+    ):
+        super().__init__(optimizer, states, hard_reset, warmup_updates)
+        if max_lr_steps <= 0:
+            raise ValueError('--lr-scheduler invsqrt requires setting --max_lr_steps')
+        self.max_lr_steps = max_lr_steps
+        self.decay_factor = np.sqrt(max(1, max_lr_steps))
+        self.scheduler = optim.lr_scheduler.LambdaLR(optimizer, self._invsqrt_lr)
+
+    def _invsqrt_lr(self, step):
+        return self.decay_factor / np.sqrt(max(1, step))
+
+    def train_step(self, scheduler_steps):
+        self.scheduler.step(epoch=scheduler_steps)
+
+    def valid_step(self, metrics_dict):
+        # this is a training step lr scheduler, nothing to adjust in validation
+        pass
+
+
+class CosineLRScheduler(ParlAILRScheduler):
+    """
+    Scheduler that decays by a cosine function.
+    """
+
+    def __init__(
+        self,
+        optimizer,
+        states,
+        hard_reset,
+        patience,
+        decay,
+        warmup_updates,
+        max_lr_steps,
+    ):
+        super().__init__(optimizer, states, hard_reset, warmup_updates)
+        if max_lr_steps <= 0:
+            raise ValueError('--lr-scheduler cosine requires setting --max_lr_steps')
+        self.max_lr_steps = max_lr_steps
+        self.scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, max_lr_steps)
+
+    def train_step(self, scheduler_steps):
+        if scheduler_steps >= self.max_lr_steps:
+            raise StopTrainException('End of Cosine LR Schedule')
+        self.scheduler.step(epoch=scheduler_steps)
+
+    def valid_step(self, metrics_dict):
+        pass
+
+
+class LinearLRScheduler(ParlAILRScheduler):
```

Comment:
@stephenroller I did a decent amount of googling and as far as I can tell PyTorch *really* doesn't do much to support Linear Decay for its schedulers, so I hacked the gamma factor, but if there's a better way to do this, I'll modify it.

Reply:
eyeballed this looks right to mean. what do you mean by gamma?

Reply:
the lr multiplier was actually wrong, returning what I believe is the right thing now.